Adopt more smart pointers in WebPageProxy (part 4)

RupinMittal opened the pull request:
#### 0d08b6b68386e4d0a7dccc51a1930faebb8b7b68
<pre>
Adopt more smart pointers in WebPageProxy (part 4)
<a href="https://bugs.webkit.org/show_bug.cgi?id=285064">https://bugs.webkit.org/show_bug.cgi?id=285064</a>
<a href="https://rdar.apple.com/141864982">rdar://141864982</a>

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/APIContentWorld.cpp:
(API::ContentWorld::pageContentWorldSingleton):
(API::ContentWorld::defaultClientWorldSingleton):
(API::ContentWorld::pageContentWorld): Deleted.
(API::ContentWorld::defaultClientWorld): Deleted.
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/C/WKUserScriptRef.cpp:
(WKUserScriptCreate):
(WKUserScriptCreateWithSource):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(+[WKContentWorld pageWorld]):
(+[WKContentWorld defaultClientWorld]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
(-[WKUserContentController addScriptMessageHandler:name:]):
(-[WKUserContentController removeScriptMessageHandlerForName:]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserScript.mm:
(-[WKUserScrip_initWithSource:injectionTime:forMainFrameOnly:includeMatchPatternStrings:excludeMatchPatternStrings:associatedURL:contentWorld:derRunningUntilNotification:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKUserStyleSheet.mm:
(-[_WKUserStyleSheet initWithSource:forMainFrameOnly:]):
(-[_WKUserStyleSheeinitWithSource:forWKWebView:forMainFrameOnly:includeMatchPatternStrings:excludeMatchPatternStrings:baseURL:level:contentWorld:]):
* Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
(webkit_user_style_sheet_new):
(webkit_user_script_new):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::~RemoteInspectorProtocolHandler):
(WebKit::RemoteInspectorProtocolHandler::handleRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::toContentWorld const):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::platformStartTask):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::sharedFileQueueSingleton):
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::WebPageProxy::tryClose):
(WebKit::WebPageProxy::setUnderPageBackgroundColorOverride):
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::receivedPolicyDecision):
(WebKit::WebPageProxy::receivedNavigationResponsePolicyDecision):
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::shouldClosePreviousPage):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::setUserAgent):
(WebKit::WebPageProxy::supportsTextEncoding const):
(WebKit::WebPageProxy::sessionState const):
(WebKit::WebPageProxy::restoreFromSessionState):
(WebKit::WebPageProxy::supportsTextZoom const):
(WebKit::WebPageProxy::setTextZoomFactor):
(WebKit::WebPageProxy::setPageZoomFactor):
(WebKit::WebPageProxy::scalePage):
(WebKit::WebPageProxy::rubberBandableEdgesRespectingHistorySwipe const):
(WebKit::WebPageProxy::findString):
(WebKit::WebPageProxy::launchInitialProcessIfNecessary):
(WebKit::WebPageProxy::runJavaScriptInMainFrame):
(WebKit::WebPageProxy::getContentsAsAttributedString):
(WebKit::WebPageProxy::saveResources):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::estimatedProgress const):
(WebKit::WebPageProxy::didStartProgress):
(WebKit::WebPageProxy::didChangeProgress):
(WebKit::WebPageProxy::didFinishProgress):
(WebKit::WebPageProxy::setNetworkRequestsInProgress):
(WebKit::WebPageProxy::updateOpener):
(WebKit::WebPageProxy::preconnectTo):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didExplicitOpenForFrame):
(WebKit::WebPageProxy::didReceiveServerRedirectForProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didChangeProvisionalURLForFrameShared):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::protectedPageLoadState):
(WebKit::WebPageProxy::protectedPageLoadState const):
(WebKit::sharedFileQueue): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/288244@main">https://commits.webkit.org/288244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/365e76926e0a85be4aaac8008ea5e5e6457583de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82233 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87365 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33294 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9777 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64087 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21835 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1364 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74822 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1264 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29004 "Build is in progress. Recent messages:") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32335 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72564 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88721 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9539 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6798 "Found 1 new test failure: http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72479 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9765 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71697 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15833 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/891 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12759 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9492 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14986 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9366 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11135 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/APIContentWorld.cpp:
```diff
@@ -88,13 +88,13 @@ Ref<ContentWorld> ContentWorld::sharedWorldWithName(const WTF::String& name, Opt
-ContentWorld& ContentWorld::pageContentWorld()
+ContentWorld& ContentWorld::pageContentWorldSingleton()
 {
     static NeverDestroyed<RefPtr<ContentWorld>> world(adoptRef(new ContentWorld(WebKit::pageContentWorldIdentifier())));
```
nit: Would be nice to use `Ref`

cdumez replied:
You misunderstood. This function should not return a smart pointer. I'd like you to use `NeverDestroyed<Ref<ContentWorld>>`. Same in other places.

RupinMittal replied:
Ah that makes sense. Fixed!

cdumez commented on Source/WebKit/UIProcess/API/APIContentWorld.cpp:
```diff
@@ -88,13 +88,13 @@ Ref<ContentWorld> ContentWorld::sharedWorldWithName(const WTF::String& name, Opt
-ContentWorld& ContentWorld::defaultClientWorld()
+ContentWorld& ContentWorld::defaultClientWorldSingleton()
 {
     static NeverDestroyed<RefPtr<ContentWorld>> world(adoptRef(new ContentWorld(WTF::String { }, { })));
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -717,7 +717,10 @@ WebPageProxy::Internals::Internals(WebPageProxy& page)
     , updateReportedMediaCaptureStateTimer(RunLoop::main(), &page, &WebPageProxy::updateReportedMediaCaptureState)
 #if HAVE(DISPLAY_LINK)
-    , wheelEventActivityHysteresis([&page](PAL::HysteresisState state) { page.wheelEventHysteresisUpdated(state); })
+    , wheelEventActivityHysteresis([weakPage = WeakPtr { &page }](PAL::HysteresisState state) {
```
No need for `&`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5120,8 +5131,8 @@ void WebPageProxy::setUserAgent(String&& userAgent, IsCustomUserAgent isCustomUs
         return;
-    forEachWebContentProcess([&](auto& webProcess, auto pageID) {
-        webProcess.send(Messages::WebPage::SetUserAgent(m_userAgent), pageID);
+    forEachWebContentProcess([userAgent = m_userAgent.isolatedCopy(), isCustomUserAgent](auto& webProcess, auto pageID) {
```
There is no multi-threading involved here, why is this calling `isolatedCopy()` now?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5369,8 +5383,11 @@ void WebPageProxy::scalePage(double scale, const IntPoint& origin, CompletionHan
     send(Messages::WebPage::DidScalePage(scale, origin));
-    forEachWebContentProcess([&] (auto& process, auto pageID) {
-        if (&process == &legacyMainFrameProcess())
+    forEachWebContentProcess([weakThis = WeakPtr { *this }, scale] (auto& process, auto pageID) {
```
This is not asynchronous, it doesn't make sense to capture a weak ptr. Just capture `page = Ref { *this }`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6298,19 +6316,22 @@ void WebPageProxy::didCreateSubframe(IPC::Connection& connection, FrameIdentifie
 
-    forEachWebContentProcess([&](auto& webProcess, auto pageID) {
-        if (!webProcess.hasConnection() || &webProcess.connection() == &connection)
+    forEachWebContentProcess([weakConnection = ThreadSafeWeakPtr { connection }, frameID](auto& webProcess, auto pageID) {
```
Not asynchronous, should not capture a weak Ptr.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6446,16 +6471,20 @@ void WebPageProxy::updateOpener(IPC::Connection& connection, WebCore::FrameIdent
         frame->updateOpener(newOpener);
-    forEachWebContentProcess([&](auto& webProcess, auto pageID) {
-        if (webProcess.hasConnection(connection))
+    forEachWebContentProcess([weakConnection = ThreadSafeWeakPtr { connection }, frameID, newOpener](auto& webProcess, auto pageID) {
```
Not async, should not capture a weak ptr.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15734,6 +15764,11 @@ void WebPageProxy::didAdjustVisibilityWithSelectors(Vector<String>&& selectors)
     m_uiClient->didAdjustVisibilityWithSelectors(*this, WTFMove(selectors));
 }
 
+Ref<BrowsingContextGroup> WebPageProxy::protectedBrowsingContextGroup() const
```
Why did this move?

RupinMittal replied:
Ah I was just moving it in the same order that the functions are listed in the header (so after `didAdjustVisibilityWithSelectors`). Do we care about the order in the cpp file?

cdumez replied:
> Do we care about the order in the cpp file?

No, so it didn't seem useful to increase patch size here.

RupinMittal replied:
Ok makes sense. Reverted.

RupinMittal committed:
Adopt more smart pointers in WebPageProxy (part 4)
https://bugs.webkit.org/show_bug.cgi?id=285064
rdar://141864982

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/APIContentWorld.cpp:
(API::ContentWorld::pageContentWorldSingleton):
(API::ContentWorld::defaultClientWorldSingleton):
(API::ContentWorld::pageContentWorld): Deleted.
(API::ContentWorld::defaultClientWorld): Deleted.
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/C/WKUserScriptRef.cpp:
(WKUserScriptCreate):
(WKUserScriptCreateWithSource):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(+[WKContentWorld pageWorld]):
(+[WKContentWorld defaultClientWorld]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
(-[WKUserContentController addScriptMessageHandler:name:]):
(-[WKUserContentController removeScriptMessageHandlerForName:]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserScript.mm:
(-[WKUserScrip_initWithSource:injectionTime:forMainFrameOnly:includeMatchPatternStrings:excludeMatchPatternStrings:associatedURL:contentWorld:derRunningUntilNotification:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKUserStyleSheet.mm:
(-[_WKUserStyleSheet initWithSource:forMainFrameOnly:]):
(-[_WKUserStyleSheeinitWithSource:forWKWebView:forMainFrameOnly:includeMatchPatternStrings:excludeMatchPatternStrings:baseURL:level:contentWorld:]):
* Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
(webkit_user_style_sheet_new):
(webkit_user_script_new):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::~RemoteInspectorProtocolHandler):
(WebKit::RemoteInspectorProtocolHandler::handleRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::toContentWorld const):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::platformStartTask):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::sharedFileQueueSingleton):
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::WebPageProxy::tryClose):
(WebKit::WebPageProxy::setUnderPageBackgroundColorOverride):
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::receivedPolicyDecision):
(WebKit::WebPageProxy::receivedNavigationResponsePolicyDecision):
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::shouldClosePreviousPage):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::setUserAgent):
(WebKit::WebPageProxy::supportsTextEncoding const):
(WebKit::WebPageProxy::sessionState const):
(WebKit::WebPageProxy::restoreFromSessionState):
(WebKit::WebPageProxy::supportsTextZoom const):
(WebKit::WebPageProxy::setTextZoomFactor):
(WebKit::WebPageProxy::setPageZoomFactor):
(WebKit::WebPageProxy::scalePage):
(WebKit::WebPageProxy::rubberBandableEdgesRespectingHistorySwipe const):
(WebKit::WebPageProxy::findString):
(WebKit::WebPageProxy::launchInitialProcessIfNecessary):
(WebKit::WebPageProxy::runJavaScriptInMainFrame):
(WebKit::WebPageProxy::getContentsAsAttributedString):
(WebKit::WebPageProxy::saveResources):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::estimatedProgress const):
(WebKit::WebPageProxy::didStartProgress):
(WebKit::WebPageProxy::didChangeProgress):
(WebKit::WebPageProxy::didFinishProgress):
(WebKit::WebPageProxy::setNetworkRequestsInProgress):
(WebKit::WebPageProxy::updateOpener):
(WebKit::WebPageProxy::preconnectTo):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didExplicitOpenForFrame):
(WebKit::WebPageProxy::didReceiveServerRedirectForProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didChangeProvisionalURLForFrameShared):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::protectedPageLoadState):
(WebKit::WebPageProxy::protectedPageLoadState const):
(WebKit::sharedFileQueue): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/288244@main
